### PR TITLE
Muzero

### PR DESCRIPTION
**Pull request by 2xic**
The code is not done yet. Some of the big things missing are

- ~~the full loss is not calculated (missing policy and value function)~~ (done)
- ~~no replay buffer -> only last game state is used currently used for testing.  (replay buffer should not be necessary for the basic environment)~~
- ~~no self play yet (should not be necerrary for the basic environment)~~
- not using residual blocks and convolution (basic change, but not needed for the basic environment)

I added a basic environment to make testing easy (see `basic_env.py` in utils). When the model is able to "win" in that environment then this PR will be ready to be merged.  

I will also update this commit with a pep8 code style checker. Seems like pylint does not report about invalid code styles.

### CODE REVIEW

File: musweeper/musweeper/muzero/tree/monte_carlo.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+		current_node.add_exploration_noise()
+		# TODO : add "or" for if current node is a terminated state 
+		while relative_depth_level < self.max_search_depth and not found_leaf:# and clock(self.timeout):
+			should_pick_random_node = self.random_rollout_metric(self, current_node)
+			current_node_has_no_children = len(current_node.children.keys()) == 0
+
+			if (should_pick_random_node or current_node_has_no_children) and len(current_node.children) != self.children_count:
+				# TODO : add something to invalidate invalid states to make search easier when bootstraping
+				new_node = current_node.get_a_children_node(self.children_count)
+				# TODO : you should actually backpropgate as soon as a leaf is found
```
Comment (2xic) [inline musweeper/musweeper/muzero/tree/monte_carlo.py]: remove comment. this is now one

File: musweeper/musweeper/muzero/tree/monte_carlo.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+		Returns
+		-------
+		node
+			leaf node after search
+		"""
+		relative_depth_level = 0
+		current_node = self.root if current_node is None else current_node
+		found_leaf = False
+		current_node.add_exploration_noise()
+		# TODO : add "or" for if current node is a terminated state
```
Comment (2xic) [inline musweeper/musweeper/muzero/tree/monte_carlo.py]: remove

File: musweeper/musweeper/muzero/tree/node.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
+import numpy as np
+import random
+
+class min_max_node_tracker:
+	def __init__(self):
+		self.max = 0
+		self.min = 0
+
+#	@staticmethod
+#	def calulate_q(self):
```
Comment (2xic) [inline musweeper/musweeper/muzero/tree/node.py]: remove, also the min - max node should be called when a node is created.

File: musweeper/musweeper/muzero/tree/node.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		dirichlet_alpha = 0.03
+		root_exploration_fraction = 0.25
+		actions = list(self.children.values())
+		noise = np.random.dirichlet([dirichlet_alpha] * len(actions))
+		for action, noise in zip(actions, noise):
+			action.prior = action.prior * (1 - root_exploration_fraction) + noise * root_exploration_fraction
+
+	def search_value_exploration_exploration(self):
+		# https://en.wikipedia.org/wiki/Monte_Carlo_tree_search
+		# TODO : Muzero uses a more complex version, change this to the one from the paper
```
Comment (2xic) [inline musweeper/musweeper/muzero/tree/node.py]: remove- this is now implemented.

UCB should be better than this function. So make UCB standard.

File: musweeper/musweeper/muzero/tree/node.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+		parrent_explored = np.log2(self.parrent.explored_count)/self.explored_count if self.parrent.explored_count != 1 and self.explored_count != 0 else 0
+		child_explored = self.wins_count / self.explored_count if self.explored_count > 0 else 0
+		c = np.sqrt(2)
+
+		return child_explored + c * np.sqrt(parrent_explored)
+
+	def upper_confidente_boundary(self):
+		self.c1 = 1.25
```
Comment (2xic) [inline musweeper/musweeper/muzero/tree/node.py]: add docstrings

File: musweeper/musweeper/muzero/utils/game_history.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+		self.actual_action = []
+
+		self.actual_value = []
+		self.value = []
+
+	@property
+	def length(self):
+		return len(self.state)
+
+class reply_buffer:
```
Comment (2xic) [inline musweeper/musweeper/muzero/utils/game_history.py]: replay

File: musweeper/musweeper/muzero/utils/game_history.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+from collections import deque
+import random
+
+class game_history:
+	def __init__(self):
```
Comment (2xic) [inline musweeper/musweeper/muzero/utils/game_history.py:7]: one of the variables are used for rollout and one for history. So you should rewrite this.

**Review by 2xic — COMMENTED** [review]
fix this tomorrow

- test the model in some other gym enviorment (cartpole or something)

- ~~since there is no "player" in the minesweeper environment. Only single player. The self play mechanism has to be based on optimal rewards in the game. Play the game x times, sort game based on total score, the lowest score is loosing players and highest score is winning player~~.  

- also the model spends a lot of time computing when retraining. We are recomputing the history when doing the rollout. I think we have to do it this way, but reread the paper and see if you missed some trick. Also, why does it take so long time when we are using cuda ?

File: musweeper/musweeper/muzero/tree/monte_carlo.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+            self.set_values_for_expand_a_node(node, model)
+
+        delta_depth = 0
+        visited_nodes = 0
+        while delta_depth < self.max_search_depth:  # and visited_nodes < 30:
+            output_node = self.expand(node, model)
+            delta_depth = (output_node.depth - node.depth)
+            self.backpropgate(output_node, delta_depth)
+            visited_nodes += 1
+        return output_node
```
Comment (Zikoat) [inline musweeper/musweeper/muzero/tree/monte_carlo.py]: The code in model.plan_action, tree.expand_node and node.expand doesnt really follow the pseudocode. 

When we run a simulation through the MCTS, we should start once at the root, then follow the path down the tree with the node that has the largest UCB and when we are at the bottom, we expand the node and backpropagate new values. We then restart the search from the top. 

It seems like what you are doing here is expanding every child of every node max_search_depth deep, but in the correct algorithm we do not care how deep down in the tree we are, we only care about how many leaf nodes we have visited.

**Pull request merged**